File: server/routes/ai.js
const express = require('express');
const { listItems } = require('../db');
const { authOptional, requireAuth, requireRole } = require('../auth');
const { enrichReport } = require('../ai');
const { smartMatches, categorizeFromText, extractColor } = require('../smartMatch');
const { analyzeImage } = require('../imageAnalysis');
const { listAdminMatches, matchToResponse } = require('../db');

const router = express.Router();

function parseQuery(body) {
  const text = String(body.query || body.description || body.name || '').trim();
  const name = String(body.name || text).trim();
  const loc = String(body.loc || '').trim();
  const description = String(body.description || text).trim();
  const type = String(body.type || 'lost').toLowerCase();
  return { name, loc, description, type };
}

router.post('/analyze-image', authOptional, async (req, res) => {
  try {
    const { photo_data, name, description, type, client_hints } = req.body || {};
    if (!photo_data || String(photo_data).length < 30) {
      return res.status(400).json({ error: 'Valid photo_data required' });
    }
    const analysis = await analyzeImage({ photo_data, name, description, type: type || 'lost', client_hints });
    res.json(analysis);
  } catch (e) {
    console.error(e);
    res.status(500).json({ error: 'Image analysis failed' });
  }
});

router.get('/monitor', requireAuth, requireRole('admin'), (_req, res) => {
  const rows = listAdminMatches()
    .filter((m) => m.status === 'pending' || m.status === 'approved')
    .slice(0, 50)
    .map((m) => {
      const r = matchToResponse(m, true);
      return {
        id: m.id,
        lost_id: m.lost_report_id,
        found_id: m.found_report_id,
        lost: r.lost_name,
        found: r.found_name,
        confidence: r.match_score,
        status: r.status,
        reason: r.match_reason,
        match: r,
      };
    });
  res.json(rows);
});

router.post('/smart-match', authOptional, (req, res) => {
  const body = req.body || {};
  const { name, loc, description, type } = parseQuery(body);
  const itemType = ['lost', 'found'].includes(type) ? type : 'lost';
  const opposite = itemType === 'lost' ? 'found' : 'lost';
  const candidates = listItems(
    (i) => i.type === opposite && ['approved', 'claimed'].includes(i.status)
  ).map((i) => ({
    ...i,
    date: new Date(i.created_at).toLocaleDateString('en-US', { month: 'short', day: 'numeric' }),
  }));
  const query = {
    name,
    loc,
    description,
    type: itemType,
    item_category: body.item_category,
    color: body.color,
    ai_tags: body.ai_tags,
    ai_description: body.ai_description,
    ai_detected_category: body.ai_detected_category,
    ai_detected_colors: body.ai_detected_colors,
    building: body.building,
    photo_data: body.photo_data,
    created_at: body.created_at,
    date_lost: body.date_lost,
  };
  const matches = smartMatches(query, candidates, 12);
  res.json({ matches, query: { name, loc, description, type: itemType } });
});

router.post('/enrich', requireAuth, requireRole('student', 'admin'), async (req, res) => {
  try {
    const { type, name, loc, description } = req.body || {};
    if (!type || !['lost', 'found'].includes(type)) {
      return res.status(400).json({ error: 'type must be lost or found' });
    }
    const out = await enrichReport({ type, draftName: name, draftLoc: loc, draftDesc: description });
    const combined = `${out.name} ${out.description} ${out.loc}`;
    res.json({
      ...out,
      item_category: categorizeFromText(combined),
      color: extractColor(combined),
    });
  } catch (e) {
    console.error(e);
    res.status(500).json({ error: 'AI enrich failed' });
  }
});

router.post('/categorize', (req, res) => {
  const text = String((req.body && req.body.text) || '').trim();
  res.json({
    item_category: categorizeFromText(text),
    color: extractColor(text),
  });
});

module.exports = router;

File: backend/routers/ai.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.deps import require_roles
from backend.models import Item, User
from backend.schemas import AIEnrichRequest, AIEnrichResponse, SmartMatchResponse
from backend.services.enrich import enrich_report
from backend.services.smart_match import smart_matches


router = APIRouter(prefix="/ai", tags=["ai"])


@router.post("/enrich", response_model=AIEnrichResponse)
async def enrich(
    payload: AIEnrichRequest,
    _user: User = Depends(require_roles("student", "admin")),
):
    item_type = payload.type.strip().lower()
    if item_type not in {"lost", "found"}:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="type must be lost or found")
    out = await enrich_report(item_type, payload.name, payload.loc, payload.description)
    return AIEnrichResponse(**out)


@router.post("/smart-match", response_model=SmartMatchResponse)
def smart_match(
    payload: AIEnrichRequest,
    _user: User = Depends(require_roles("student", "admin")),
    db: Session = Depends(get_db),
):
    item_type = payload.type.strip().lower()
    if item_type not in {"lost", "found"}:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="type must be lost or found")

    opposite = "found" if item_type == "lost" else "lost"
    candidates = db.scalars(
        select(Item).where(Item.type == opposite, Item.status.in_(["approved", "claimed"])).order_by(Item.created_at.desc())
    ).all()
    matches = smart_matches(
        {"name": payload.name, "loc": payload.loc, "description": payload.description},
        candidates,
    )
    return SmartMatchResponse(matches=matches)
